### PR TITLE
[ESI][Windows] Lower Window ports to unions

### PR DESCRIPTION
**Pull request by teqdruid**
Introduces a new pass for lowering ESI types to something synthesizable. Currently, that's just data windows to unions.

### CODE REVIEW

File: lib/Dialect/ESI/ESIPasses.cpp [hunk 1]
```diff
@@ -1388,6 +1388,169 @@ void ESIPortsPass::updateInstance(HWModuleExternOp mod, InstanceOp inst) {
...
+  if (!window)
+    return false;
+  hw::UnionType &lowered = typeCache[window];
+  if (!lowered)
+    lowered = window.getLoweredType();
+  port.type = lowered;
+  return true;
+}
+
+void ESITypesPass::lowerMod(HWMutableModuleLike mod) {
```
Comment (mortbopet) [inline lib/Dialect/ESI/ESIPasses.cpp:1428]: hmm... I know this is probably the "CIRCT-canonical" way of doing module-level type modifications (mainly due to `mod.modifyPorts`). I'm wondering whether you've also (now or at an earlier point) experimented with using the conversion infrastructure?
It seems applicable since you're basically DIYing a type converter + symbol cache + type conversion materialization.

**Review by mortbopet — APPROVED** [review]
As commented, this seems like something that _may_ fit into the type conversion infrastructure. I'd like to see an attempt at that, but I'm not going to make you rewrite all of this code (although most of it could probably be repurposed with only a slight diff). Feel free to merge if you don't want to give it a shot 👍 .

Secondly, nit - i look forward to the day that ESIPasses.cpp is split into separate files... 2325 lines of many many passes now. Personally, that was one of my pet peeves when working with ESI passes - when i want to navigate to a specific pass I'd open the file that was equally named as the pass, and i'd directly know where to go in that file to find what I'm looking for. This is very much not possible with ESIPasses.cpp.

**Comment by teqdruid** [comment]
> As commented, this seems like something that _may_ fit into the type conversion infrastructure

Yeah, I realized that 1/2 way through... only problem is that I don't fully understand the type conversion infrastructure!

**Comment by mortbopet** [comment]
Playing devil's advocate here; seeing as there isn't (afaik) a rush to get this in, this might be a good opportunity to learn!

**Comment by teqdruid** [comment]
> > Secondly, nit - i look forward to the day that ESIPasses.cpp is split into separate files... 2325 lines of many many passes now. Personally, that was one of my pet peeves when working with ESI passes - when i want to navigate to a specific pass I'd open the file that was equally named as the pass, and i'd directly know where to go in that file to find what I'm looking for. This is very much not possible with ESIPasses.cpp.
> 
> Yeah, I agree. This has been on my list for some time now.

I moved lower types into its own file. Will do the same for the other ones after this.

**Comment by teqdruid** [comment]
@mortbopet I've learned and put as much as makes sense into the dialect/type conversion framework. I'm not using the block signature conversion stuff on modules since I can't use it to handle external modules. It also doesn't handle terminators. So

File: lib/Dialect/ESI/Passes/ESILowerTypes.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  for (auto port : ports.outputs) {
+    if (!lowerPort(mod, port))
+      continue;
+    loweredOutputs.emplace_back(port.argNum, port);
+    loweredOutputIdxs.push_back(port.argNum);
+    if (!body)
+      continue;
+
+    // Build an unwrap to unwrap the window into the lowered output. This will
+    // hopefully be canonicalized away.
```
Comment (mortbopet) [inline lib/Dialect/ESI/Passes/ESILowerTypes.cpp:103]: Shouldn't target materialization take care of this?

File: lib/Dialect/ESI/Passes/ESILowerTypes.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  for (auto port : ports.inputs) {
+    if (!lowerPort(mod, port))
+      continue;
+    loweredInputs.emplace_back(port.argNum, port);
+    loweredInputIdxs.push_back(port.argNum);
+    if (!body)
+      continue;
+
+    // Build a wrapper op to wrap the lowered value. This will hopefully be
+    // canonicalized away.
```
Comment (mortbopet) [inline lib/Dialect/ESI/Passes/ESILowerTypes.cpp:80]: Shouldn't source materialization take care of this?

**Review by mortbopet — COMMENTED** [review]
Nice!

I actually suspect that you can leverage the type conversion infra even more.
I think the HWArithToHW pattern is a great example of this (which essentially needs to do exactly the same as this pass - lower `hw.module` types). There, all of the `hw.module` finickyness implicitly handled by having a fully generic pattern to convert all ops which contain the offending types (https://github.com/llvm/circt/blob/main/lib/Conversion/HWArithToHW/HWArithToHW.cpp#L352-L362). (with only a small [addition](https://github.com/llvm/circt/blob/main/lib/Conversion/HWArithToHW/HWArithToHW.cpp#L391-L392) needed to handle the annoying arg attribute stuff).

Then, [the entire program](https://github.com/llvm/circt/blob/main/lib/Conversion/HWArithToHW/HWArithToHW.cpp#L505-L506) is marked as having to need full conversion (all ops must be legalized or converted) and  legality is defined [much like you already define legality in this version of your pass](https://github.com/llvm/circt/blob/main/lib/Conversion/HWArithToHW/HWArithToHW.cpp#L131).

so tl;dr: if you copy the [TypeConversionPattern](https://github.com/llvm/circt/blob/main/lib/Conversion/HWArithToHW/HWArithToHW.cpp#L391-L392) verbatim (maybe we should just have that somewhere in `Support` since it seems generically useful), and assuming your source/target materializers work correctly, you should be able to remove both the `InstanceConversionPattern` and `ModuleConversionPattern`.

**Comment by teqdruid** [comment]
@mortbopet Yep -- `TypeConversionPattern` worked _nearly_ out of the box. I had to waste some time figuring out why it was crashing without realizing it was crashing only with the '-debug' flag. It's actually what I was looking for -- something where the conversion was entirely defined by the type converter. I wonder if we should upstream it?

**Comment by mortbopet** [comment]
I suspect that you're encountering https://github.com/llvm/circt/issues/4899.

> I wonder if we should upstream it?

I think it would be nice to have upstream (i mean, i grabbed the pattern from SPIRV, anyways, and only modified it slightly wrt. the function type and argument attribute copying).

**Review by mortbopet — APPROVED** [review]
